src/betree.c

```diff
@@ -0,0 +1,1147 @@
+#include <math.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <ctype.h>
+
+#include "ast.h"
+#include "betree.h"
+
+bool match_sub(const struct event* event, const struct sub *sub)
+{
+    if(sub == NULL) {
+        return false;
+    }
+    return match_node(event, sub->expr) == 1;
+} 
+
+void check_sub(const struct event* event, const struct lnode* lnode, struct matched_subs* matched_subs) 
+{
+    for(unsigned int i=0; i < lnode->sub_count; i++) {
+        const struct sub* sub = lnode->subs[i];
+        if(match_sub(event, sub) == true) {
+            if(matched_subs->sub_count == 0) {
+                matched_subs->subs = malloc(sizeof(int));
```

Comment:
The general pattern I use for allocating memory:
`matched_subs->subs = calloc(1, sizeof(matched_subs->subs));`

The benefits of this pattern are:
1) use calloc instead of malloc as calloc initialises the memory to 0 which is generally a good idea.
2) the `sizeof` is calculated on the variable being allocated instead of the type of that variable. The advantage here is that if you change the type of the variable and you won't have to remember to change the sizeof malloc statement.